### ISSUE TITLE
twisted.python.reflect.safe_str on Python 3 should convert utf-8 encoded bytes to clean str

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7660]
|<img alt="wolfgang61's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| wolfgang61 reported|
|-|-|
|Trac ID|trac#7660|
|Type|defect|
|Created|2014-10-03 11:10:25Z|
|Branch|https://github.com/twisted/twisted/tree/utf8-safe-str-7660|

safe_str(b'a') now returns "b'a'", wanted is "a"

Attachments:

* [7660a.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/449/4496237cf04a7b6fd47009d971a4f736318d48ea/8466e2abde8468bc11557af2ea5007a6beda497a.patch) (3284 bytes) - added by wolfgang61 on 2014-10-03 11:12:56Z - 
* [7660b.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/449/4496237cf04a7b6fd47009d971a4f736318d48ea/64eb4c7483997e2da6e73e67cd9367a4ca962646.patch) (3792 bytes) - added by wolfgang61 on 2014-10-18 03:58:24Z - 
* [7660c.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/449/4496237cf04a7b6fd47009d971a4f736318d48ea/c1ac69170d89015e03e86523a153169e1cf1606e.patch) (4372 bytes) - added by wolfgang61 on 2014-10-18 04:04:39Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7660 7660
type__defect defect
reporter__wolfgang61 wolfgang61
priority__normal normal
milestone__ 
branch__branches_utf8_safe_str_7660 branches/utf8-safe-str-7660
branch_author__jml jml
status__closed closed
resolution__fixed fixed
component__core core
keywords__spread spread
time__1412334625548789 1412334625548789
changetime__1474198049187875 1474198049187875
version__None None
owner__jml jml

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="wolfgang61's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|wolfgang61 commented|
|-|-|

Otherwise Failure.getErrorMessage() for Python 3 will return "b'Error message'" instead of "Error Message". This breaks applications which rightly expect a native str from getErrorMessage() for direct display to the user.

Since the return type for getErrorMessage() is undefined by docstring, I guess I can expect str.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @jml|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_4"></a>|@jml commented|
|-|-|

(In [43399]) Branching to utf8-safe-str-7660.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_5"></a>|@jml commented|
|-|-|

(In [43400]) Initial commit of wolfgang61's patch refs [#7660](https://github.com/twisted/twisted/issues/7660)

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_6"></a>|@jml set owner to wolfgang61|
|-|-|

Thanks!

I've made a lot of comments, but they are mostly pretty simple.

1. For future reference, the NEWS file for t.python should go in `twisted/topfiles`. I've done it for you in the branch – no action required.
1. Please update the comments in `_safeFormat` to be capitalized sentences.
1. Also, the comment in `_safeFormat` isn't very clear. Perhaps you should also say what you _do_ want?
1. [OPTIONAL] I wonder if `_safeFormat` would be better deleted, with the `except` block being moved to an internal helper function, and the `try` block being inlined into `safe_repr` and `safe_str` with the necessary changes.
1. [OPTIONAL] In `test_workingRepr`, the `for` loop is OK, but it means that if the first one fails we don't learn anything about the behaviour of the second. I'd be tempted to do something like this:
```
working_objects = [[1, 2, 3], b'a']
self.assertEqual(map(safe_repr, working_objects), map(repr, working_objects))
```
1. `test_reflect:546`: Comment should begin with a string
1. Also, that comment is unclear. What is legal UTF8? What do we want to be a native string? What _is_ a native string?
1. The docstrings for `test_workingUtf8_2` and `test_workingUtf8_3` seem to have weirdly quoted strings. In particular, the string `b'utf8-data, Python 2'` isn't used anywhere in the test. I think a better docstring would say "safe_str turns bytes b'a' into a str in Python 2", or similar.
1. Likewise for `test_brokenUtf8`. How does including `"b'non-utf8'"` help the docstring?
1. Both `test_workingUtf8_2` and `test_workingUtf8_3` would probably benefit from being split, with the second example (`b't\xc3\xbcst'`) going into a second test. That way, the test can be given a name that describes what's interesting about that example.

Please fix the non-optional comments and assign back to me (or someone else!) for review.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_7"></a>|@jml commented|
|-|-|

Sorry, forgot these other lint warnings:

```
W9402:546,0: The first letter of comment should be capitalized
W9015:560,0: Too many blank lines, found 2
C0301:567,0: Line too long (81/79)
```

**Comment by twisted-trac** [comment]
|<img alt="wolfgang61's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|wolfgang61 set owner to @jml|
|-|-|

Please apply only 7660c.patch on branch utf8-safe-str-7660.

Normally I would have made several commits of this, but I believe you prefer it this way.

I cannot run twistedchecker anymore. I am now on pylint 1.3.1, and twistedchecker insists on pylint 0.26.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_9"></a>|@jml commented|
|-|-|

Thanks.

I made the following changes to get it ready for landing: r43407, r43408, r43409, r43410.

For reference:

* Long line in `test_reflect`
* The `map` thing I told you was a lie. `map` returns an iterator in Python 3, so it needs to be converted to a list to do equality comparison.
* The `working_objects` name I gave you was also a lie. In Twisted, it should be `workingObjects`. I've renamed to `xs`, since the name isn't very important in this context.
* `@return` for `_safeFormat`

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_10"></a>|@jml set status to `closed`|
|-|-|

(In [43411]) safe_str now converts UTF8 bytes to str in Python 3

* Author: wolfgang61
* Reviewer: jml
* Fixes: [#7660](https://github.com/twisted/twisted/issues/7660)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl commented|
|-|-|

[mass edit] Removing review from closed tickets.
